validator/src/main.rs

```diff
@@ -256,9 +267,16 @@ fn get_rpc_node(
             match highest_snapshot_hash {
                 None => {
                     assert!(eligible_rpc_peers.is_empty());
-                    info!("No snapshots available");
                 }
                 Some(highest_snapshot_hash) => {
+                    if eligible_rpc_peers.is_empty() {
+                        retry_reason = Some(format!(
+                            "Wait for newer snapshot than local: {:?}",
+                            highest_snapshot_hash
+                        ));
+                        continue;
+                    }
```

Comment:
:meat_on_bone:

Reply:
Nit: I this log message doesn't account for
1) `trusted_snapshot_hashes` is empty: https://github.com/solana-labs/solana/blob/master/validator/src/main.rs#L221-L222  or  
2)`rpc_peers` is empty: https://github.com/solana-labs/solana/blob/master/validator/src/main.rs#L181-L185

Reply:
If local has a newer snapshot why don't we just use it.

We can do this generically when trusted validators exist:  if all available snapshots from trusted validators are older than what local, use local.   There could even be a new cli argument, `--use-local-snapshot-if-newer-than <NUMBER_OF_SLOTS>` that would use a local snapshot if it's at most `NUMBER_OF_SLOTS` slots behind the newest trusted validator.  Default value of `NUMBER_OF_SLOTS` is maybe 100-500.

Reply:
@carllin 

> Nit: I this log message doesn't account for:

Well, I misread the code too but these `continue`s are for the inner loop: https://github.com/solana-labs/solana/blob/master/validator/src/main.rs#L226

Reply:
> If local has a newer snapshot why don't we just use it.

The main reason is that **original snapshot RPC selection logic doesn't wait for the full set of specified rpc trusted nodes to be available**. So, we can't determine that predicate to begin with.

So, if a booting validator sees a subset of trusted rpc nodes with older snapshot, it considers the snapshot to be the newest one the cluster has to offer and fetches it.


Definitely, we can wait to determine ` if all available snapshots from trusted validators are older than what local, ...`, but this requires 100% liveness of trusted validators. Given two or three trusted nodes for mainnet-beta/testnet, using 90% or 80% liveness threshold doesn't quite work well.

Also, it's operationally rare to have newer local snapshot assuming solid setup uses `--no-snapshot-fetch` anyway and/or the connected cluster aren't making newer snapshots.

So this new change is rather for corner-casing. Also it's for development purpose, where frequent and forcible restarts are rampant. ;)

So, I opted for simplicity.

Reply:
Also, I'll consider `--use-local-snapshot-if-newer-than` for a bit...

Reply:
> `--use-local-snapshot-if-newer-than <NUMBER_OF_SLOTS>` that would use a local snapshot if it's at most `NUMBER_OF_SLOTS` slots behind the newest trusted validator. Default value of `NUMBER_OF_SLOTS` is maybe 100-500.

Well, sadly supporting this takes a bit of effort than just plumbling (reasons below). How about merging this pr first? So that v1.4 (=~ testnet) doesn't suffer this known persistent-tower bug? (Heh, I was completely unaware of this snapshot-fetch-related failure scenario...).

The detailed reason of more efforts:

First, persistent tower never supports older root bank than voted banks to prevent double vote and other general slashing events. So, we can't use local snapshots even if it's just slightly older (practically, like 100-500) while voting is enabled. 

Persistent tower always assumes local availability of voted slots which aren't isolated like snapshot's root banks. The rationale is that you should have the slots and have replayed them if you voted on them (duh).

The only exception is to start over newer snapshots. This is for some disk failure or corrupted rocksdb.

Reply:
@mvines Could you check my previous comment?

Especially,

> Well, sadly supporting this takes a bit of effort than just plumbling (reasons below). How about merging this pr first? So that v1.4 (=~ testnet) doesn't suffer this known persistent-tower bug? (Heh, I was completely unaware of this snapshot-fetch-related failure scenario...).

Reply:
Works for me.  This gets us is more in the right direction